Add state-in-constructor rule

lukyth opened the pull request:
This will resolve #1810. Comments are very welcomed. 😃

alexzherdev commented on lib/rules/state-in-constructor.js:
```diff
@@ -0,0 +1,59 @@
+          option === 'never' &&
+          node.left.object.type === 'ThisExpression' &&
+          node.left.property.name === 'state' &&
+          node.parent.parent.parent.parent.kind === 'constructor' &&
```
This is a bit brittle: assignment could be nested in other nodes (e.g. if conditions) in which case this sequence will not work. This should either be traversing parents in a loop, or scopes like https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/no-unused-prop-types.js#L264 (although not sure which one is better)

ljharb replied:
Let's try to add a test case that would fail here.

alexzherdev commented on lib/rules/state-in-constructor.js:
```diff
@@ -0,0 +1,59 @@
+          node.left.object.type === 'ThisExpression' &&
+          node.left.property.name === 'state' &&
+          node.parent.parent.parent.parent.kind === 'constructor' &&
+          utils.isES6Component(node.parent.parent.parent.parent.parent.parent)
```
Same here, but in this case maybe `utils.getParentES6Component` would work

ljharb commented on lib/rules/state-in-constructor.js:
```diff
@@ -0,0 +1,59 @@
+      AssignmentExpression(node) {
+        if (
+          option === 'never' &&
+          node.left.object.type === 'ThisExpression' &&
```
Let's try to add a test case that would fail here.

lukyth commented:
@alexzherdev @ljharb Thank you so much for your reviews! I've addressed all of them with these commits 81d0e8b...b1024c0. Could you please check them out?

I also removed some test cases which I think are redundant (81d0e8b). I'm not sure if that's the right call. If you guys don't think so then I can bring them back.

krawaller commented:
Just to inject some energy, this looks sweet! Good work @lukyth! :)

lukyth commented:
Anything I can do to get this merged?

lukyth commented:
@EvHaus I've added this rule to the list in `README.md` (9415814). Sorry it took me so long. I just saw your comment here.
